### PR TITLE
Added question in array folder Maximum_Sum_sub_squre_matrix

### LINKED ISSUE #845: WANT TO ADD QUESTION IN ARRAYS Folder "Maximum Sum Square SubMatrix : Given an N x N matrix, find a k x k submatrix where k <= N and k >= 1, such that sum of all the elements in submatrix is maximum. The input matrix can contain zero, positive and negative numbers"
---

Hi @ajay-dhangar,

**Subject**: Implementation of Maximum Sum Square SubMatrix using an Optimized Approach

**Description**:

This issue has been addressed by providing a solution to the problem of finding the submatrix with the maximum sum in a given 2D matrix of size N x N. The goal is to find the maximum sum of all possible square submatrices of a specified size within the matrix.

**Key Points**:

- **Problem Description**: The task is to compute the maximum sum of a square submatrix of a given size in the input matrix. The solution efficiently calculates the sum of all possible submatrices and returns the maximum sum.
  
- **Solution Approach**: The algorithm iterates through the matrix while calculating prefix sums, allowing for efficient calculation of the sum of any square submatrix in constant time. This optimized approach ensures that the problem is solved in a time-efficient manner, significantly reducing complexity.

Thank you for your time and consideration!

Please assign me to this task. It would be a valuable opportunity for me to contribute and further develop my problem-solving skills.

I look forward to hearing from you!

Best regards,  
@khurshed07  
https://github.com/khurshed07  

---

### PR DESCRIPTION
**Pull request by khurshed07**
issue #845

### Description
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #845

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# Implementation of MAXIMUM SUM OF SUB-SQUARE MATRIX for Issue #845

## Description
This pull request addresses Issue #845 by implementing the solution to the problem of finding the maximum sum of a k x k sub-square matrix within a given n x n matrix. The objective is to efficiently compute the sum of all possible k x k sub-squares and identify the one with the highest sum.

## Approach
The solution is implemented using a two-step approach:

1. **Preprocessing Strip Sums:** The algorithm calculates the sums of all strips of size k x 1 for each column in the matrix. This preprocessing step allows for efficient computation of k x k sub-square sums.

2. **Sum Calculation of Sub-squares:** Using the precomputed strip sums, the algorithm calculates the sum of each k x k sub-square and updates the maximum sum found.

## Algorithm
- **Construct Strip Sums:** An auxiliary array is created to store the sums of strips of size k x 1 for each column.
- **Calculate Sum of Sub-squares:** Iterate over the matrix to compute the sums of all k x k sub-squares using the precomputed strip sums.
- **Update Maximum Sum:** Keep track of the maximum sum and its position in the matrix.

## Code Explanation
- **printMaxSumSub Function:** The main function where the logic for finding the maximum sum sub-square matrix is implemented. It handles preprocessing and computes the maximum sum of the k x k sub-square.
- **Matrix Declaration and Driver Code:** The driver code demonstrates how to use the function with an example matrix.

## Files Added
- **Max_Sum_Sub_Square_Matrix.md:** Contains the problem statement, solution explanation, and steps to run the program.
- **Max_Sum_Sub_Square_Matrix.cpp:** The C++ program implementing the algorithm to solve the problem.

### The file has been placed in the Array folder.

---

Thank you for your time and consideration!

Best regards,  
@khurshed07  
[GitHub Profile](https://github.com/khurshed07)

### CODE REVIEW

File: arrays/Maximum Sum Sub-square Matrix.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        { 4, 4, 4, 4, 4 },
+        { 5, 5, 5, 5, 5 },
+    };
+    int k = 3;
+
+    cout << "Maximum sum 3 x 3 matrix is\n";
+    printMaxSumSub(mat, k);
+
+    return 0;
+}
```
Comment (ajay-dhangar) [inline arrays/Maximum Sum Sub-square Matrix.md]: close your example with ```` ``` ````

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
update your code based on suggestions

**Comment by khurshed07** [comment]
Hi @ajay-dhangar,

I have closed the example with ``` as required. Could you please check it and submit the pull request? If there are any issues, feel free to let me know.

Thank you!

@khurshed07 khurshed

**Review by ajay-dhangar — APPROVED** [review]
Done

**Pull request merged**